File: app/src/main/java/com/example/fastwork/utils/value/ValueUtil.java
package com.example.fastwork.utils.value;

import android.content.Context;
import android.util.Log;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Map;
import java.util.Stack;
import java.util.UUID;

public class ValueUtil {
    private static final boolean DEBUG=true;

    //    private static ThreadLocal<Long> sTime=new ThreadLocal<>();
    private static ThreadLocal<Stack<Long>> sTimes = new ThreadLocal<>();

    static {
        sTimes.set(new Stack<Long>());
    }

    public static SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
    public static SimpleDateFormat sSimpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
    public static SimpleDateFormat timeFormat = new SimpleDateFormat("HH:mm:ss");

    public static String formatDate(Date date) {
        return dateFormat.format(date);
    }

    public static String formatDate(long date) {
        return dateFormat.format(date);
    }

    public static String formatTime(long date) {
        return timeFormat.format(date);
    }

    public static int compare(int lhs,int rhs){
        return lhs < rhs ? -1 : (lhs == rhs ? 0 : 1);
    }
    public static int compare(long lhs,long rhs){
        return lhs < rhs ? -1 : (lhs == rhs ? 0 : 1);
    }
    public static Date parseDate(String date) {
        if (date == null) return null;
        try {
            return dateFormat.parse(date);
        } catch (ParseException e) {
            Log.e("ValueUtil", "parseDate", e);
            return null;
        }
    }

    public static Integer toInteger(Object value) {
        Double v = toDouble(value);
        if (v != null) return v.intValue();
        return null;
    }

    public static Double toDouble(Object value) {
        if (value == null) return null;
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        try {
            return Double.parseDouble(value.toString());
        } catch (NumberFormatException e) {

        }
        return null;
    }


    public static String getStr(Map<String, Object> map, String key, String def) {
        Object value = map.get(key);
        if (value == null) return def;
        return value.toString();
    }

    public static int getInt(Map<String, Object> map, String key, int def) {
        Object value = map.get(key);
        if (value == null) return def;
        Integer ret = toInteger(value);
        return ret == null ? def : ret;
    }

    public static double getFloat(Map<String, Object> map, String key, double def) {
        Object value = map.get(key);
        if (value == null) return def;
        Double ret = toDouble(value);
        return ret == null ? def : ret;
    }


    public static boolean getBool(Map<String, Object> map, String key, boolean def) {
        Object value = map.get(key);
        if (value == null) return def;
        return Boolean.parseBoolean(value.toString());
    }

    public static int dip2px(Context context, float dipValue) {

        float scale = context != null ? context.getResources().getDisplayMetrics().density : 1;
        return (int) (dipValue * scale + 0.5f);
    }






    public static String uuid() {
        return UUID.randomUUID().toString().replaceAll("\\-", "");
    }

    public static int parseInt(String s) {
        return parseInt(s, 0);
    }

    public static int parseInt(String s, int def) {
        try {
            return Integer.parseInt(s);
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
        return def;
    }

    public static float parseFloat(String name) {
        try {
            return Float.parseFloat(name);
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}

File: app/src/main/java/com/example/fastwork/utils/value/ValueUtil.kt
package com.example.fastwork.utils.value

import android.app.ActivityManager
import android.content.Context
import android.support.multidex.BuildConfig
import com.example.fastwork.utils.log.Lg

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Stack
import java.util.UUID


object ValueUtil {
    private val DEBUG = true

    //    private static ThreadLocal<Long> sTime=new ThreadLocal<>();
    private val sTimes = ThreadLocal<Stack<Long>>()

    var dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    var sSimpleDateFormat = SimpleDateFormat("yyyy-MM-dd")
    var timeFormat = SimpleDateFormat("HH:mm:ss")

    init {
        sTimes.set(Stack())
    }

    fun formatDate(date: Date): String {
        return dateFormat.format(date)
    }

    fun formatDate(date: Long): String {
        return dateFormat.format(date)
    }

    fun formatTime(date: Long): String {
        return timeFormat.format(date)
    }

    fun compare(lhs: Int, rhs: Int): Int {
        return if (lhs < rhs) -1 else if (lhs == rhs) 0 else 1
    }

    fun compare(lhs: Long, rhs: Long): Int {
        return if (lhs < rhs) -1 else if (lhs == rhs) 0 else 1
    }

    fun parseDate(date: String?): Date? {
        if (date == null) return null
        try {
            return dateFormat.parse(date)
        } catch (e: ParseException) {
            //            Lg.e("ValueUtil", "parseDate", e);
            return null
        }

    }

    fun toInteger(value: Any): Int? {
        val v = toDouble(value)
        return v?.toInt()
    }

    fun toDouble(value: Any?): Double? {
        if (value == null) return null
        if (value is Number) {
            return value.toDouble()
        }
        try {
            return java.lang.Double.parseDouble(value.toString())
        } catch (e: NumberFormatException) {

        }

        return null
    }


    fun getStr(map: Map<String, Any>, key: String, def: String): String {
        val value = map[key] ?: return def
        return value.toString()
    }

    fun getInt(map: Map<String, Any>, key: String, def: Int): Int {
        val value = map[key] ?: return def
        val ret = toInteger(value)
        return ret ?: def
    }

    fun getFloat(map: Map<String, Any>, key: String, def: Double): Double {
        val value = map[key] ?: return def
        val ret = toDouble(value)
        return ret ?: def
    }


    fun getBool(map: Map<String, Any>, key: String, def: Boolean): Boolean {
        val value = map[key] ?: return def
        return java.lang.Boolean.parseBoolean(value.toString())
    }


    fun dip2px(context: Context?, dipValue: Float): Int {
        var scale = context?.resources?.displayMetrics?.density ?: 1 as Float
        return (dipValue * scale + 0.5f).toInt()
    }


    fun startTime() {
        if (!BuildConfig.DEBUG) return

        //        sTime.set(System.nanoTime());
        var stack = sTimes.get()
        if (stack == null) {
            sTimes.set(Stack<Long>())

        }
        stack!!.push(System.nanoTime())
    }

    fun endTime(msg: String?): Float {
        if (!BuildConfig.DEBUG) return 0f
        val stack = sTimes.get()
            ?: //            Lg.w("ValueUtil", "endTime:please invoke startTime before endTime");
            return -1f
        val start = stack.pop() ?: return 0f
        val use = (System.nanoTime() - start) / 1000000f
        if (msg != null) {
                        Lg.i("endTime", msg + " : " + use + "ms");
        }
        if (stack.size == 0) {
            sTimes.set(null)
        }
        return use
    }

    fun uuid(): String {
        return UUID.randomUUID().toString().replace("\\-".toRegex(), "")
    }

    @JvmOverloads
    fun parseInt(s: String, def: Int = 0): Int {
        try {
            return Integer.parseInt(s)
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }

        return def
    }

    fun parseFloat(name: String): Float {
        try {
            return java.lang.Float.parseFloat(name)
        } catch (e: NumberFormatException) {
            return 0f
        }

    }
}
